File: app/src/main/java/com/ternaryop/photoshelf/service/CounterIntentService.java
package com.ternaryop.photoshelf.service;

import java.util.Calendar;

import android.app.IntentService;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.ternaryop.photoshelf.db.DBHelper;
import com.ternaryop.photoshelf.event.CounterEvent;
import com.ternaryop.tumblr.Tumblr;
import com.ternaryop.tumblr.TumblrUtils;
import org.greenrobot.eventbus.EventBus;

/**
 * Created by dave on 28/10/17.
 * Handle counters retrieval
 */
public class CounterIntentService extends IntentService implements PhotoShelfIntentExtra {
    public CounterIntentService() {
        super("counterIntent");
    }

    @Override
    protected void onHandleIntent(@Nullable Intent intent) {
        if (intent == null) {
            return;
        }
        String selectedBlogName = intent.getStringExtra(EXTRA_BLOG_NAME);
        int type = intent.getIntExtra(EXTRA_TYPE, CounterEvent.NONE);
        String action = intent.getStringExtra(EXTRA_ACTION);

        if (ACTION_FETCH_COUNTER.equals(action)) {
            fetchCounterByType(type, selectedBlogName);
        }
    }

    public static void fetchCounter(@NonNull Context context,
                                    @NonNull String blogName,
                                    @CounterEvent.CounterType int type) {
        Intent intent = new Intent(context, CounterIntentService.class);
        intent.putExtra(EXTRA_TYPE, type);
        intent.putExtra(EXTRA_BLOG_NAME, blogName);
        intent.putExtra(EXTRA_ACTION, ACTION_FETCH_COUNTER);

        context.startService(intent);
    }

    private void fetchCounterByType(final @CounterEvent.CounterType int type, final String blogName) {
        new Thread(new Runnable() {
            @Override
            public void run() {
                if (EventBus.getDefault().hasSubscriberForEvent(CounterEvent.class)) {
                    EventBus.getDefault().post(new CounterEvent(type, getCount(type, blogName)));
                }
            }
        }).start();
    }

    private long getCount(final @CounterEvent.CounterType int type, final String blogName) {
        try {
            switch (type) {
                case CounterEvent.BIRTHDAY:
                    return DBHelper
                            .getInstance(getApplicationContext())
                            .getBirthdayDAO()
                            .getBirthdaysCountInDate(Calendar.getInstance().getTime(), blogName);
                case CounterEvent.DRAFT:
                    return TumblrUtils.getDraftCount(Tumblr.getSharedTumblr(getApplicationContext()), blogName);
                case CounterEvent.SCHEDULE:
                    return TumblrUtils.getQueueCount(Tumblr.getSharedTumblr(getApplicationContext()), blogName);
                case CounterEvent.NONE:
                    break;
            }
        } catch (Exception ignored) {
        }
        return 0;
    }
}

File: app/src/main/java/com/ternaryop/photoshelf/service/CounterIntentService.kt
package com.ternaryop.photoshelf.service

import android.app.IntentService
import android.content.Context
import android.content.Intent
import com.ternaryop.photoshelf.EXTRA_BLOG_NAME
import com.ternaryop.photoshelf.EXTRA_TYPE
import com.ternaryop.photoshelf.api.ApiManager
import com.ternaryop.photoshelf.api.birthday.FindParams
import com.ternaryop.photoshelf.event.CounterEvent
import com.ternaryop.tumblr.android.TumblrManager
import com.ternaryop.tumblr.draftCount
import com.ternaryop.tumblr.queueCount
import com.ternaryop.utils.date.dayOfMonth
import com.ternaryop.utils.date.month
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import org.greenrobot.eventbus.EventBus
import java.util.Calendar

/**
 * Created by dave on 28/10/17.
 * Handle counters retrieval
 */
class CounterIntentService : IntentService("counterIntent") {

    override fun onHandleIntent(intent: Intent?) {
        intent ?: return

        val selectedBlogName = intent.getStringExtra(EXTRA_BLOG_NAME) ?: return
        val type = intent.getIntExtra(EXTRA_TYPE, CounterEvent.NONE)

        if (ACTION_FETCH_COUNTER == intent.action) {
            fetchCounterByType(type, selectedBlogName)
        }
    }

    private fun fetchCounterByType(type: Int, blogName: String) {
        Thread(Runnable {
            if (EventBus.getDefault().hasSubscriberForEvent(CounterEvent::class.java)) {
                EventBus.getDefault().post(CounterEvent(type, getCount(type, blogName)))
            }
        }).start()
    }

    private fun getCount(type: Int, blogName: String): Int {
        try {
            return when (type) {
                CounterEvent.BIRTHDAY -> birthdayCount()
                CounterEvent.DRAFT -> TumblrManager.getInstance(applicationContext).draftCount(blogName)
                CounterEvent.SCHEDULE -> TumblrManager.getInstance(applicationContext).queueCount(blogName)
                CounterEvent.NONE -> 0
                else -> -1
            }
        } catch (ignored: Exception) {
        }

        return 0
    }

    private fun birthdayCount(): Int = runBlocking(Dispatchers.IO) {
        val now = Calendar.getInstance()
        ApiManager.birthdayService().findByDate(
            FindParams(onlyTotal = true, month = now.month + 1, dayOfMonth = now.dayOfMonth).toQueryMap())
            .response.total.toInt()
    }

    companion object {
        private const val ACTION_FETCH_COUNTER = "fetchCounter"

        fun fetchCounter(context: Context,
                         blogName: String,
                         type: Int) {
            val intent = Intent(context, CounterIntentService::class.java)
                .setAction(ACTION_FETCH_COUNTER)
                .putExtra(EXTRA_TYPE, type)
                .putExtra(EXTRA_BLOG_NAME, blogName)

            context.startService(intent)
        }
    }
}
